src/ApiCore/ResourceTemplate/RelativeResourceTemplate.php

```diff
@@ -0,0 +1,296 @@
+<?php
+/*
+ * Copyright 2018 Google LLC
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Google\ApiCore\ResourceTemplate;
+
+use Google\ApiCore\ValidationException;
+
+/**
+ * Represents a relative resource template, meaning that it will never contain a leading slash or
+ * trailing verb (":<verb>").
+ *
+ * Examples:
+ *   projects
+ *   projects/{project}
+ *   foo/{bar=**}/fizz/*
+ *
+ * Templates use the syntax of the API platform; see
+ * https://github.com/googleapis/api-common-protos/blob/master/google/api/http.proto
+ * for details. A template consists of a sequence of literals, wildcards, and variable bindings,
+ * where each binding can have a sub-path. A string representation can be parsed into an
+ * instance of AbsoluteResourceTemplate, which can then be used to perform matching and instantiation.
+ */
+class RelativeResourceTemplate implements ResourceTemplateInterface
+{
+    /** @var Segment[] $segments */
+    private $segments;
+
+    /**
+     * RelativeResourceTemplate constructor.
+     *
+     * @param string $path
+     * @throws ValidationException
+     */
+    public function __construct($path)
+    {
+        if (empty($path)) {
+            throw new ValidationException("Cannot construct RelativeResourceTemplate from empty string");
+        }
+        $this->segments = Parser::parseSegments($path);
+
+        $doubleWildcardCount = self::countDoubleWildcards($this->segments);
```

Comment:
Are there use cases where we would want a list of segments that have more than 1 double wildcard? If not, do you think it would be worthwhile to track the wildcards and throw an exception within `Parser::parseSegments()` as we're building the list?

Reply:
I think I would rather leave `parseSegments` as just parsing unconnected segments which must be valid on their own, and then apply check that apply to all segments (wildcards, duplicate keys) afterwards. WDYT?